x/metadata/types/signer_utils.go

```diff
@@ -0,0 +1,429 @@
+package types
+
+import (
+	"bytes"
+	"fmt"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/authz"
+)
+
+// PartyDetails is a struct used to help process party and signer validation.
+// Even though all the fields are public, you should usually use the Get/Set methods
+// which handle automatic bech32 conversion when needed and reduce duplicated efforts.
+type PartyDetails struct {
+	// address is the bech32 account address string of this party.
+	address string
+	// role is the type of this party.
+	role PartyType
+	// optional indicates whether this party's signature is required (false => required, true => not required).
+	optional bool
+
+	// acc is the converted address field.
+	acc sdk.AccAddress
+
+	// signer is the bech32 string of the account signing for this party.
+	// It might be different from the address field if authz is involved.
+	signer string
+	// signerAcc is the account that's signing for this party.
+	// It might be different from the acc field if authz is involved.
+	signerAcc sdk.AccAddress
+
+	// canBeUsedBySpec indicates whether this party can be used to fulfill a role required by the spec.
+	canBeUsedBySpec bool
+	// usedBySpec indicates whether this party has been used to fulfill a role required by the spec.
+	usedBySpec bool
+}
+
+// WrapRequiredParty creates a PartyDetails from the provided Party.
+func WrapRequiredParty(party Party) *PartyDetails {
+	return &PartyDetails{
+		address:  party.Address,
+		role:     party.Role,
+		optional: party.Optional,
+	}
+}
+
+// WrapAvailableParty creates a PartyDetails from the provided Party and marks it as optional and usable.
+func WrapAvailableParty(party Party) *PartyDetails {
+	return &PartyDetails{
+		address:         party.Address,
+		role:            party.Role,
+		optional:        true, // An available party is optional unless something else says otherwise.
+		canBeUsedBySpec: true,
+	}
+}
+
+// BuildPartyDetails creates the list of PartyDetails to be used in party/signer/role validation.
+func BuildPartyDetails(reqParties, availableParties []Party) []*PartyDetails {
+	details := make([]*PartyDetails, 0, len(availableParties))
+
+	// Start with creating details for each available party.
+availablePartiesLoop:
+	for _, party := range availableParties {
+		for _, known := range details {
+			if party.IsSameAs(known) {
+				continue availablePartiesLoop
+			}
+		}
+		details = append(details, WrapAvailableParty(party))
+	}
+
+	// Now update the details to include optional=false required parties.
+	// If an equal party is already in the details, just update its optional flag
+	// to false, otherwise, add it to the list.
+reqPartiesLoop:
```

Comment:
Same as above. There is probably a way to simplify this using the slices package.

Reply:
Good catch. It'll be updated with the next push.